platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/LongListDisk.java

```diff
@@ -125,20 +257,51 @@ public synchronized boolean putIfEqual(final long index, final long oldValue, fi
     @Override
     public void writeToFile(final Path newFile) throws IOException {
         // finish writing to current file
-        fileChannel.force(true);
+        currentFileChannel.force(true);
         // if new file is provided then copy to it
-        if (!file.equals(newFile)) {
-            try (final FileChannel fc =
-                    FileChannel.open(newFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
-                fileChannel.position(0);
-                MerkleDbFileUtils.completelyTransferFrom(fc, fileChannel, 0, fileChannel.size());
-            }
+        try (final FileChannel fc = FileChannel.open(newFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+            // write header
+            writeHeader(fc);
+            // write data
+            writeLongsData(fc);
+            fc.force(true);
         }
     }
 
-    /** No-op as we override writeToFile directly */
+    /**
+     * {@inheritDoc}
+     */
     @Override
-    protected void writeLongsData(final FileChannel fc) {}
+    protected void writeLongsData(final FileChannel fc) throws IOException {
+        final int totalNumOfChunks = calculateNumberOfChunks(size());
+        final long currentMinValidIndex = minValidIndex.get();
+        final int firstChunkWithDataIndex = (int) currentMinValidIndex / numLongsPerChunk;
+        final ByteBuffer transferBuffer = transferBufferThreadLocal.get();
+        for (int i = firstChunkWithDataIndex; i < totalNumOfChunks; i++) {
+            final Long chunkOffset;
+            if (i == firstChunkWithDataIndex) {
+                // writing starts from the first valid index in the first valid chunk
+                final int firstValidIndexInChunk = (int) currentMinValidIndex % numLongsPerChunk;
+                transferBuffer.position(firstValidIndexInChunk * Long.BYTES);
+                chunkOffset = chunkList.get(i) + calculateOffsetInChunk(currentMinValidIndex);
+            } else {
+                transferBuffer.position(0);
+                chunkOffset = chunkList.get(i);
+            }
+            if (i == (totalNumOfChunks - 1)) {
+                // last array, so set limit to only the data needed
+                final long bytesWrittenSoFar = (long) memoryChunkSize * (long) i;
+                final long remainingBytes = (size() * Long.BYTES) - bytesWrittenSoFar;
+                transferBuffer.limit((int) remainingBytes);
+            } else {
+                transferBuffer.limit(transferBuffer.capacity());
+            }
+            int currentPosition = transferBuffer.position();
+            MerkleDbFileUtils.completelyRead(currentFileChannel, transferBuffer, chunkOffset);
+            transferBuffer.position(currentPosition);
+            MerkleDbFileUtils.completelyWrite(fc, transferBuffer);
```

Comment:
I wonder if `MerkleDbFileUtils.completelyTransferFrom()` can be used here?

Reply:
Please take a look at the latest implementation. I think combination of  `completelyRead` and `completelyWrite` make it a bit easier to reason about what's going on there.